tests/asgi/tests.py

```diff
@@ -194,6 +197,56 @@ async def test_post_body(self):
         self.assertEqual(response_body["type"], "http.response.body")
         self.assertEqual(response_body["body"], b"Echo!")
 
+    async def test_cancel_post_request_with_sync_processing(self):
+        loop = asyncio.get_event_loop()
+        # Events used to monitor the view processing from the parent test code
+        view_started_event = asyncio.Event()
+        view_finished_event = asyncio.Event()
+        # Record exception raised in the test view
+        exceptions = []
+
+        # This view will run in a new thread because it is wrapped in sync_to_async.
+        # The view consumes the POST body data after a short delay. The test will cancel
+        # the request using http.disconnect during the delay, but because this is a sync
+        # view the code runs to completion. There should be no exceptions raised inside
+        # the view code.
+        @csrf_exempt
+        @sync_to_async
+        def echo_view(request):
+            try:
+                loop.call_soon_threadsafe(view_started_event.set)
+                time.sleep(0.1)
```

Comment:
I wonder if we could wait on a `threading.Event` here, rather than sleeping 🤔

Reply:
I've tried a few things here but I haven't managed to do this with events and keep the test behaviour that I want (the test should fail if I move the body_file.close() in ASGIHandler to before the cleanup code). If using an Event for synchronisation, either the view code gets unblocked too early (and the test doesn't test anything) or I try to put it later and get a deadlock and the test never finishes. While imperfect this might be the only way to get anything tested without making the synchronisation code very complex.

Reply:
One thing I would change in this code is having something like an `outcome = []` list to append mappings with the exception and/or body, so when later asserting on its content, the test failure is more informative. Specifically, I would suggest this diff:

```diff
diff --git a/tests/asgi/tests.py b/tests/asgi/tests.py
index 955ec884d4..b562a64c8c 100644
--- a/tests/asgi/tests.py
+++ b/tests/asgi/tests.py
@@ -206,10 +206,8 @@ class ASGITest(SimpleTestCase):
         # Events used to monitor the view processing from the parent test code
         view_started_event = asyncio.Event()
         view_finished_event = asyncio.Event()
-        # Record exception raised in the test view
-        exceptions = []
-        # Request body
-        request_body = []
+        # Record received request body or exceptions raised in the test view
+        outcome = []
 
         # This view will run in a new thread because it is wrapped in
         # sync_to_async. The view consumes the POST body data after a short
@@ -224,10 +222,10 @@ class ASGITest(SimpleTestCase):
                 loop.call_soon_threadsafe(view_started_event.set)
                 time.sleep(0.1)
                 # Do something to read request.body after pause
-                request_body.append(request.body)
+                outcome.append({"request_body": request.body})
                 return HttpResponse("ok")
             except Exception as e:
-                exceptions.append(e)
+                outcome.append({"exception": e})
             finally:
                 loop.call_soon_threadsafe(view_finished_event.set)
 
@@ -257,8 +255,7 @@ class ASGITest(SimpleTestCase):
         with self.assertRaises(asyncio.TimeoutError):
             await communicator.receive_output()
 
-        self.assertEqual(request_body, [b"Body data!"])
-        self.assertEqual(len(exceptions), 0)
+        self.assertEqual(outcome, [{"request_body": b"Body data!"}])
 
     async def test_untouched_request_body_gets_closed(self):
         application = get_asgi_application()
```

That on error produces this failure:

```
AssertionError: Lists differ: [{'exception': ValueError('I/O operation on closed file.')}] != [{'request_body': b'Body data!'}]

First differing element 0:
{'exception': ValueError('I/O operation on closed file.')}
{'request_body': b'Body data!'}

- [{'exception': ValueError('I/O operation on closed file.')}]
+ [{'request_body': b'Body data!'}]
```

Instead of:

```
AssertionError: Lists differ: [] != [b'Body data!']

Second list contains 1 additional elements.
First extra element 0:
b'Body data!'

- []
+ [b'Body data!']
```